x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// GetCode implements the `GetCode` method for the `Authenticator` interface
+func (a CustomAuthenticator) GetCode() uint8 {
+	return statute.MethodUserPassAuth
+}
+
+// Authenticate checks if the username matches a required prefix and extracts the suffix
+func (a CustomAuthenticator) Authenticate(reader io.Reader, writer io.Writer, userAddr string) (*socks5.AuthContext, error) {
+	// Respond to the client to use username/password authentication
+	if _, err := writer.Write([]byte{statute.VersionSocks5, statute.MethodUserPassAuth}); err != nil {
+		return nil, err
+	}
+
+	// Parse the username and password from the client request
+	nup, err := statute.ParseUserPassRequest(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	fullUsername := string(nup.User)
+	password := string(nup.Pass)
+
+	// Split the username into prefix and suffix parts
+	parts := strings.SplitN(fullUsername, "-", 2)
+	if len(parts) < 2 {
+		if _, err := writer.Write([]byte{statute.UserPassAuthVersion, statute.AuthFailure}); err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("username does not contain a valid prefix-suffix format")
+	}
+
+	usernamePrefix := parts[0] // Prefix (e.g., "jane")
+	usernameSuffix := parts[1] // Suffix (e.g., "-country-ir")
+
+	log.Printf("Attempting authentication with prefix: %s and suffix: %s", usernamePrefix, usernameSuffix)
+
+	// Validate the prefix and password against stored credentials
+	if !a.PrefixCredentials.Valid(usernamePrefix, password, userAddr) {
+		// Authentication failed
+		if _, err := writer.Write([]byte{statute.UserPassAuthVersion, statute.AuthFailure}); err != nil {
+			return nil, err
+		}
+		return nil, statute.ErrUserAuthFailed
+	}
+
+	// Authentication successful
+	if _, err := writer.Write([]byte{statute.UserPassAuthVersion, statute.AuthSuccess}); err != nil {
+		return nil, err
+	}
+
+	// Return the context with the extracted suffix for later use
+	return &socks5.AuthContext{
+		Method: statute.MethodUserPassAuth,
+		Payload: map[string]string{
+			"suffix": usernameSuffix, // Store the suffix for later use in the dialer
+			"prefix": usernamePrefix, // Store the prefix as well, if needed
+		},
+	}, nil
+}
+
+// CredentialStore is an interface to validate prefix-based credentials
+type CredentialStore interface {
+	Valid(username, password, userAddr string) bool
+}
+
+// StaticCredentialStore is a simple credential store that holds a map of valid prefixes and passwords
+type StaticCredentialStore struct {
+	credentials map[string]string // Maps username/prefix to password
+}
+
+// Valid checks if the provided prefix and password are valid
+func (s *StaticCredentialStore) Valid(prefix, password, userAddr string) bool {
+	storedPassword, exists := s.credentials[prefix]
+	return exists && storedPassword == password
+}
+
+func udpAssociateHandler(ctx context.Context, writer io.Writer, request *socks5.Request) error {
```

Comment:
Let's remove this and handle UDP in the Server dialer instead.

I'm not comfortable exposing the internals of our SOCKS5 library for now and that will require some more thought.

Using the dialer approach will keep the code simpler, even though it's doing an extra relay of the packets.

Reply:
@fortuna this is an important part of the implementation. I am basically implementing a custom `udpAssociateHandler` which relays the UDP associate command coming from the client (to proxy server) to the upstream soax server and relays the bind address communicated by soax server back to the client. 

I initially thought I can use `WithDialAndRequest` to relay UDP as well but this handler only handle outgoing TCP connections (and not even the UDP associate TCP connection). I guess you are suggesting that I use a packetDialer similar to SteamDialer in the `WithDialAndRequest` for `udp` networks right? If so, unfortunately that approach won't work since `WithDialAndRequest` won't be called for `udp`.

Reply:
The library already implements the relay:
https://github.com/things-go/go-socks5/blob/503e7866821a419fe3766751f6d5671390072c14/handle.go#L219

The dial to the target happens here:
https://github.com/things-go/go-socks5/blob/503e7866821a419fe3766751f6d5671390072c14/handle.go#L242

It's directly from `Server.dial`:
https://github.com/things-go/go-socks5/blob/503e7866821a419fe3766751f6d5671390072c14/handle.go#L183

Which is set by `WithDialAndRequest`. That applies to both tcp and udp.

You will need to take the credentials from the request, create the SOCKS5 dialer, then dial.

Reply:
Ah no! It doesn't actually use `Server.dialWithRequest`!
So, yeah, we need to provide the relay.